File: scripts/cd.jl
using DelimitedFiles
using LinearAlgebra
using Formatting
using PyPlot

function CD_calc(dir)
  mu_file   = format("{}{}", dir, "mu_site.out")
  com_file  = format("{}{}", dir, "c_o_m.dat")
  eig_file  = format("{}{}", dir, "eigvecs.out")
  en_file   = format("{}{}", dir, "eigvals.out")
  mu        = readdlm(mu_file)
  r         = readdlm(com_file)
  eig       = readdlm(eig_file)
  en        = readdlm(en_file)
  ns = 2048
  nchl = 14
  λ = 1.0
  CD = zeros(ns)
  wn = zeros(ns)
  g_sum = zeros(ns)
  chi = Array{Real}(undef, nchl, ns, 2)
  for k = 1:nchl
    chi_file    = format("{}{}{:02d}{}", dir, "chi_i_", k, ".dat")
    chi[k, :, :]      = readdlm(chi_file) 
  end
  wn = chi[1, :, 1]

  for n = 1:nchl
    for m = 1:nchl
      dd = ((mu[n, :]) × (mu[m, :])) ⋅ (r[n, :] - r[m, :])
      g_sum = zeros(ns)
      for k = 1:nchl
        # chi_file = format("{}{}{:02d}{}", dir, "chi_i_", k, ".dat")
        # chi      = readdlm(chi_file) 
        # if (n == 1 && m == 1 && k == 1)
        #   wn       = chi[:, 1]
        # end
        # s        = sum(chi[:, 2])Cross product - W
        g        = chi[k, :, 2] / sum(chi[k, :, 2])
        g_sum .+= (g .* (eig[k, n] * eig[k, m] * en[k]))
      end
      CD .+= (g_sum .* dd)
    end
  end

  CD .*= (-16 * pi^2 / 9λ) .* wn
  # CD .*= (-16 * pi^2 / 9λ) .* wn
  wl = replace!(1e7 ./ wn, Inf=>NaN)

  open(format("{}{}", dir, "CD.dat"), "w") do io
    writedlm(io, [wl CD])
  end

  return [wl wn CD]
end

dir = ARGS[1]
if (dir[end] != "/")
  dir *= "/"
end
res = CD_calc(dir)

# p = plot(res[:, 1], res[:, 3], xaxis = ("frequency (nm)", (580, 720)))
p = plot(res[:, 1], res[:, 3])
ax = gca()
ax.set_xlim([580, 720])
xlabel("frequency (nm)")
ylabel("CD")
savefig(format("{}{}", dir, "CD.pdf"))

File: spectra/src/cd.c
#include "cd.h"

void
cross(double *a, double* b, double *res)
{
  res[0] = a[1] * b[2] - a[2] * b[1];
  res[1] = a[2] * b[0] - a[0] * b[2];
  res[2] = a[0] * b[1] - a[1] * b[0];
}

void
dot(double *a, double* b, double res)
{
  res = (a[0] * b[0]) + (a[1] * b[1]) + (a[2] * b[2]);
}

void
cd(unsigned n_chl, unsigned ns, double **chiw, double **mu,
   double **eig, double **com, double *eigvals, double *cd)
{
  double v[3], rnm[3];
  double dd = 0;
  double *chiw_sums = (double *)calloc(n_chl, sizeof(double));
  /* double *g = (double *)calloc(ns, sizeof(double)); */
  for (unsigned n = 0; n < n_chl; n++) {
    double sum = 0.;
    for (unsigned m = 0; m < ns; m++) {
      sum += chiw[n][m];
    }
    chiw_sums[n] = sum;
  }


  for (unsigned n = 0; n < n_chl; n++) {
    for (unsigned m = 0; m < n_chl; m++) {
      cross(mu[n], mu[m], v);
      for (unsigned l = 0; l < 3; l++) {
        rnm[l] = com[m][l] = com[n][l];
      }
      dot(v, rnm, dd);

      for (unsigned k = 0; k < n_chl; k++) {
        for (unsigned l = 0; l < ns; l++) {
          cd[l] += (2 * PI * ((float)l / ns) * dd)
                * ((chiw[k][l] * eig[k][m] * 
                  eig[k][n] * eigvals[k])
                / chiw_sums[k]);
        }
      }

    }
  }
}
